### PR TITLE
LLM: bigdl-llm lora support & lora example

### PR DESCRIPTION
**Pull request by rnwang04**
### 1. Why the change?

Provide better support of lora on Intel GPUs, and accelerate the training speed.
benchmark and verification is here : https://github.com/analytics-zoo/nano/issues/878#issuecomment-1866091540

### 2. User API changes

See example.

### 3. Summary of the change 

- add `BF16Linear` & `LoraBF16Linear` class to accelerate bf16 lora finetuning
- add cast function to save memory and accelerate the training speed
- provide single card example of lora, and local verification
- provide multi card example of lora, and local verification
- refactor code of lora and qlora

### 4. How to test?
- [x] Unit test
- [x] Local test of llama2-7b lora finetuning on single tile PVC
- [x] Local test of llama2-7b lora finetuning on multi card PVC

### CODE REVIEW

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/alpaca_qlora_finetuning.py [hunk 1]
```diff
@@ -109,6 +108,7 @@ def train(
     prompt_template_name: str = "alpaca",  # The prompt template to use, will default to alpaca.
     gradient_checkpointing: bool = False,
     deepspeed: str = None,
+    lora: bool = False, # if True, use lora
     qa_lora: bool = False, # if True, use qa-lora https://arxiv.org/abs/2309.14717
```
Comment (jason-dai) [inline python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/alpaca_qlora_finetuning.py:112]: Combine into something like `training_mode: str = "qlora"`

File: python/llm/src/bigdl/llm/transformers/qlora.py [hunk 2]
```diff
@@ -157,7 +234,7 @@ def _create_new_module(create_new_module_func, lora_config, adapter_name, target
 
 @dataclass
 class LoraConfig(LoraConfigBase):
-
+    lora: bool = field(default=False, metadata={"help": "enable lora"})
     qa_lora: bool = field(default=False, metadata={"help": "enable qa-lora"})
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/qlora.py:238]: combine these two into a something like `mode: str =(default="qlora", ...)`

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/alpaca_qlora_finetuning.py (hunk 1 above)
Comment (rnwang04) [reply python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/alpaca_qlora_finetuning.py]: Done.

File: python/llm/src/bigdl/llm/transformers/qlora.py (hunk 2 above)
Comment (rnwang04) [reply python/llm/src/bigdl/llm/transformers/qlora.py]: Have changed to `training_mode: str` here.

File: python/llm/src/bigdl/llm/transformers/qlora.py [hunk 3]
```diff
@@ -234,8 +230,7 @@ def _create_new_module(create_new_module_func, lora_config, adapter_name, target
 
 @dataclass
 class LoraConfig(LoraConfigBase):
-    lora: bool = field(default=False, metadata={"help": "enable lora"})
-    qa_lora: bool = field(default=False, metadata={"help": "enable qa-lora"})
+    training_mode: str = field(default="qlora", metadata={"help": "determine training mode"})
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/qlora.py:233]: add more info in help: what are possible values, what is the default value

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/alpaca_qlora_finetuning.py [hunk 4]
```diff
@@ -176,9 +178,9 @@ def train(
...
         # According to the QLoRA paper, using "nf4" could yield better model quality than "int4"
         # Default 4-bit format for qa-lora is sym_int4
-        if qa_lora:
+        if training_mode == "qalora":
             low_bit_format = "sym_int4"
-        elif lora:
+        elif training_mode == "lora":
             low_bit_format = "bf16"
         else:
             low_bit_format = "nf4"
```
Comment (jason-dai) [inline python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/alpaca_qlora_finetuning.py]: check if `training_mode` is valid

Comment (rnwang04) [reply python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/alpaca_qlora_finetuning.py]: Have added check in line 115-116 of this file:
```python
invalidInputError(training_mode in ["qlora", "qalora", "lora"],
                  "Only qlora / qalora / lora are supported for training_mode now.")
```

File: python/llm/src/bigdl/llm/transformers/qlora.py (hunk 3 above)
Comment (rnwang04) [reply python/llm/src/bigdl/llm/transformers/qlora.py]: Done.

**Comment by jason-dai** [comment]
Please update example readme to clearly describe the different modes that we support: lora, qlora, qa_lora. Otherwise LGTM

**Comment by rnwang04** [comment]
> Please update example readme to clearly describe the different modes that we support: lora, qlora, qa_lora. Otherwise LGTM

Have simply update this part. For further example improvement, maybe can be done in https://github.com/intel-analytics/BigDL/pull/9719

**Review by cyita — APPROVED** [review]
LGTM

File: python/llm/src/bigdl/llm/transformers/qlora.py [hunk 5]
```diff
@@ -128,22 +129,94 @@ def forward(self, x: torch.Tensor):
...
-                                      target.in_features,
-                                      target.out_features,
-                                      bias=bias,
-                                      **low_bit_kwargs)
+
+        if lora_config.training_mode != "lora":
+            low_bit_kwargs.update(
+                {
+                    "qtype": target.qtype,
+                    "qa_lora": True if lora_config.training_mode == "qalora" else False,
```
Comment (cyita) [inline python/llm/src/bigdl/llm/transformers/qlora.py:206]: The peft config may not have attr `training_mode`.

Comment (rnwang04) [reply python/llm/src/bigdl/llm/transformers/qlora.py]: Have fixed this.

**Comment by rnwang04** [comment]
Failed to run lora on Arc770 as OOM, which seems a little unreasonable.
Will merge this PR first, and take a look at above issue later.

**Pull request merged**